src/decoder/mod.rs

```diff
@@ -205,34 +208,39 @@ fn decode_bson<R: Read + ?Sized>(reader: &mut R, tag: u8, utf8_lossy: bool) -> D
             reader.take(len as u64).read_to_end(&mut data)?;
             Ok(Bson::Binary(subtype, data))
         }
-        Some(ObjectId) => {
+        Some(ElementType::ObjectId) => {
             let mut objid = [0; 12];
             for x in &mut objid {
                 *x = reader.read_u8()?;
             }
             Ok(Bson::ObjectId(oid::ObjectId::with_bytes(objid)))
         }
-        Some(Boolean) => Ok(Bson::Boolean(reader.read_u8()? != 0)),
-        Some(NullValue) => Ok(Bson::Null),
-        Some(RegularExpression) => {
+        Some(ElementType::Boolean) => Ok(Bson::Boolean(reader.read_u8()? != 0)),
+        Some(ElementType::NullValue) => Ok(Bson::Null),
+        Some(ElementType::RegularExpression) => {
             let pattern = read_cstring(reader)?;
             let options = read_cstring(reader)?;
             Ok(Bson::RegExp(RegExp { pattern, options }))
         }
-        Some(JavaScriptCode) => read_string(reader, utf8_lossy).map(Bson::JavaScriptCode),
-        Some(JavaScriptCodeWithScope) => {
+        Some(ElementType::JavaScriptCode) => {
+            read_string(reader, utf8_lossy).map(Bson::JavaScriptCode)
+        }
+        Some(ElementType::JavaScriptCodeWithScope) => {
             // disregard the length:
             //     using Read::take causes infinite type recursion
             read_i32(reader)?;
 
             let code = read_string(reader, utf8_lossy)?;
             let scope = decode_document(reader)?;
-            Ok(Bson::JavaScriptCodeWithScope(code, scope))
+            Ok(Bson::JavaScriptCodeWithScope(JavaScriptCodeWithScope {
```

Comment:
This is the only change to method besides adding the `ElementType::` prefix to the cases